chain/store/store.go

```diff
@@ -779,6 +856,47 @@ func (cs *ChainStore) persistCompactHeight(height uint32) error {
 	return cs.st.Put(db.TrieCompactHeightKey(), heightBuffer)
 }
 
+func (cs *ChainStore) persistSyncRootHeight(height uint32) error {
+	heightBuffer := make([]byte, 4)
+	binary.LittleEndian.PutUint32(heightBuffer[:], height)
+	return cs.st.Put(db.TrieSyncRootHeightKey(), heightBuffer)
+}
+
+func (cs *ChainStore) persistSyncStatus(status uint32) error {
+	statusBuffer := make([]byte, 4)
+	binary.LittleEndian.PutUint32(statusBuffer[:], status)
+	return cs.st.Put(db.TrieFastSyncStatusKey(), statusBuffer)
+}
+
+func (cs *ChainStore) getSyncRootHeight() uint32 {
+	heightBuffer, err := cs.st.Get(db.TrieSyncRootHeightKey())
+	if err != nil {
+		return 0
+	}
+	return binary.LittleEndian.Uint32(heightBuffer)
+}
+
+func (cs *ChainStore) GetSyncRootHeight() uint32 {
+	return cs.getSyncRootHeight()
+}
+
+func (cs *ChainStore) getFastSyncStatus() uint32 {
+	statusBuffer, err := cs.st.Get(db.TrieFastSyncStatusKey())
+	if err != nil {
+		log.Info("no previous fast sync status found")
+		return beforeFastSync
+	}
+	return binary.LittleEndian.Uint32(statusBuffer)
+}
+
+func (cs *ChainStore) GetFastSyncStatus() uint32 {
+	return cs.getFastSyncStatus()
+}
+
+func (cs *ChainStore) FastSyncRestorable() bool {
```

Comment:
Change this function to `ShouldFastSync` and change logic to:
if state is beforeFastSync, return true only if local block height == 0
if state is beginFastSync, return true
if state is postFastSync, return false